File: DesEcbStringEncryption/DesEcbStringEncryption.java
package DES_ECB_String;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.Base64;


public class DesEcbStringEncryption {
    public static void main(String[] args) throws NoSuchAlgorithmException, IllegalBlockSizeException, InvalidKeyException, BadPaddingException, InvalidAlgorithmParameterException, NoSuchPaddingException {
        System.out.println("\n# # # SECURITY WARNING: This code is provided for achieve    # # #");
        System.out.println("# # # compatibility between different programming languages. # # #");
        System.out.println("# # # It is not necessarily fully secure.                    # # #");
        System.out.println("# # # It uses the OUTDATED and UNSECURE algorithm DES.       # # #");
        System.out.println("# # # It uses FIXED key that should NEVER used,              # # #");
        System.out.println("# # # instead use random generated keys.                     # # #");
        System.out.println("# # # DO NOT USE THIS CODE IN PRODUCTION                     # # #\n");

        String plaintext = "The quick brown fox jumps over the lazy dog";
        System.out.println("plaintext:  " + plaintext);

        System.out.println("\nDES ECB String encryption with fixed key");

        // fixed encryption key 8 bytes long
        byte[] desEncryptionKey = "12345678".getBytes(StandardCharsets.UTF_8);
        System.out.println("desEncryptionKey (Base64): " + base64Encoding(desEncryptionKey));

        // encryption
        System.out.println("\n* * * Encryption * * *");
        String desCiphertextBase64 = desEcbEncryptToBase64(desEncryptionKey, plaintext);
        System.out.println("ciphertext: " + desCiphertextBase64);
        System.out.println("output is (Base64) ciphertext");
        // decryption
        System.out.println("\n* * * Decryption * * *");
        System.out.println("ciphertext (Base64): " + desCiphertextBase64);
        System.out.println("input is (Base64) ciphertext");
        String decryptedtext = desEcbDecryptFromBase64(desEncryptionKey, desCiphertextBase64);
        System.out.println("plaintext:  " + decryptedtext);

        System.out.println("\nTriple DES ECB String encryption with fixed key (16 bytes)");

        // fixed encryption key 16 bytes long
        byte[] des2EncryptionKey = "1234567890123456".getBytes(StandardCharsets.UTF_8);
        System.out.println("des2EncryptionKey (Base64): " + base64Encoding(des2EncryptionKey));

        // encryption
        System.out.println("\n* * * Encryption * * *");
        String des2CiphertextBase64 = des3EcbEncryptToBase64(tdes2ToDes3Key(des2EncryptionKey), plaintext);
        System.out.println("ciphertext: " + des2CiphertextBase64);
        System.out.println("output is (Base64) ciphertext");
        // decryption
        System.out.println("\n* * * Decryption * * *");
        System.out.println("ciphertext (Base64): " + des2CiphertextBase64);
        System.out.println("input is (Base64) ciphertext");
        decryptedtext = des3EcbDecryptFromBase64(tdes2ToDes3Key(des2EncryptionKey), des2CiphertextBase64);
        System.out.println("plaintext:  " + decryptedtext);

        System.out.println("\nTriple DES ECB String encryption with fixed key (24 bytes)");

        // fixed encryption key 24 bytes long
        byte[] des3EncryptionKey = "123456789012345678901234".getBytes(StandardCharsets.UTF_8);
        System.out.println("des3EncryptionKey (Base64): " + base64Encoding(des3EncryptionKey));

        // encryption
        System.out.println("\n* * * Encryption * * *");
        String des3CiphertextBase64 = des3EcbEncryptToBase64(des3EncryptionKey, plaintext);
        System.out.println("ciphertext: " + des3CiphertextBase64);
        System.out.println("output is (Base64) ciphertext");
        // decryption
        System.out.println("\n* * * Decryption * * *");
        System.out.println("ciphertext (Base64): " + des3CiphertextBase64);
        System.out.println("input is (Base64) ciphertext");
        decryptedtext = des3EcbDecryptFromBase64(des3EncryptionKey, des3CiphertextBase64);
        System.out.println("plaintext:  " + decryptedtext);
    }

    private static byte[] generateRandomDesKey() {
        SecureRandom secureRandom = new SecureRandom();
        byte[] key = new byte[8]; // 8 bytes for des, 16 bytes for TDES 2 and 24 bytes for TDES 3
        secureRandom.nextBytes(key);
        return key;
    }

    private static String desEcbEncryptToBase64(byte[] desKey, String data) throws NoSuchPaddingException, NoSuchAlgorithmException,
            InvalidKeyException, BadPaddingException, IllegalBlockSizeException {
        SecretKeySpec secretKeySpec = new SecretKeySpec(desKey, "DES");
        Cipher cipher = Cipher.getInstance("DES/ECB/PKCS5Padding");
        cipher.init(Cipher.ENCRYPT_MODE, secretKeySpec);
        String ciphertextBase64 = base64Encoding(cipher.doFinal(data.getBytes(StandardCharsets.UTF_8)));
        return ciphertextBase64;
    }

    private static String desEcbDecryptFromBase64(byte[] desKey, String data) throws NoSuchPaddingException, NoSuchAlgorithmException, InvalidAlgorithmParameterException, InvalidKeyException, BadPaddingException, IllegalBlockSizeException {
        byte[] encryptedData = base64Decoding(data);
        Cipher cipher = Cipher.getInstance("DES/ECB/PKCS5Padding");
        SecretKeySpec secretKeySpec = new SecretKeySpec(desKey, "DES");
        cipher.init(Cipher.DECRYPT_MODE, secretKeySpec);
        return new String(cipher.doFinal(encryptedData));
    }

    private static String des3EcbEncryptToBase64(byte[] des3Key, String data) throws NoSuchPaddingException, NoSuchAlgorithmException,
            InvalidKeyException, BadPaddingException, IllegalBlockSizeException {
        SecretKeySpec secretKeySpec = new SecretKeySpec(des3Key, "DESede");
        Cipher cipher = Cipher.getInstance("DESede/ECB/PKCS5Padding");
        cipher.init(Cipher.ENCRYPT_MODE, secretKeySpec);
        String ciphertextBase64 = base64Encoding(cipher.doFinal(data.getBytes(StandardCharsets.UTF_8)));
        return ciphertextBase64;
    }

    private static String des3EcbDecryptFromBase64(byte[] des3Key, String data) throws NoSuchPaddingException, NoSuchAlgorithmException, InvalidAlgorithmParameterException, InvalidKeyException, BadPaddingException, IllegalBlockSizeException {
        byte[] encryptedData = base64Decoding(data);
        Cipher cipher = Cipher.getInstance("DESede/ECB/PKCS5Padding");
        SecretKeySpec secretKeySpec = new SecretKeySpec(des3Key, "DESede");
        cipher.init(Cipher.DECRYPT_MODE, secretKeySpec);
        return new String(cipher.doFinal(encryptedData));
    }

    private static byte[] tdes2ToDes3Key(byte[] tdes2Key) {
        byte[] tdes3Key = new byte[24];
        System.arraycopy(tdes2Key, 0, tdes3Key, 0, 16);
        System.arraycopy(tdes2Key, 0, tdes3Key, 16, 8);
        return tdes3Key;
    }

    private static String base64Encoding(byte[] input) {
        return Base64.getEncoder().encodeToString(input);
    }

    private static byte[] base64Decoding(String input) {
        return Base64.getDecoder().decode(input);
    }
}

File: DesEcbStringEncryption/DesEcbStringEncryption.php
<?php
function generateRandomDesKey()
{
    return openssl_random_pseudo_bytes(8, $crypto_strong);
}

function base64Encoding($input)
{
    return base64_encode($input);
}

function base64Decoding($input)
{
    return base64_decode($input);
}

function desEcbEncryptToBase64($key, $data)
{
    $ciphertext = openssl_encrypt($data, 'des-ecb', $key, OPENSSL_RAW_DATA);
    return base64_encode($ciphertext);
}

function desEcbDecryptFromBase64($key, $data)
{
    return openssl_decrypt(base64_decode($data), 'des-ecb', $key, OPENSSL_RAW_DATA);
}

function des2EcbEncryptToBase64($key, $data)
{
    $ciphertext = openssl_encrypt($data, 'des-ede-ecb', $key, OPENSSL_RAW_DATA);
    return base64_encode($ciphertext);
}

function des2EcbDecryptFromBase64($key, $data)
{
    return openssl_decrypt(base64_decode($data), 'des-ede-ecb', $key, OPENSSL_RAW_DATA);
}

function des3EcbEncryptToBase64($key, $data)
{
    $ciphertext = openssl_encrypt($data, 'des-ede3-ecb', $key, OPENSSL_RAW_DATA);
    return base64_encode($ciphertext);
}

function des3EcbDecryptFromBase64($key, $data)
{
    return openssl_decrypt(base64_decode($data), 'des-ede3-ecb', $key, OPENSSL_RAW_DATA);
}

echo PHP_EOL . '# # # SECURITY WARNING: This code is provided for achieve    # # #' . PHP_EOL;
echo '# # # compatibility between different programming languages. # # #' . PHP_EOL;
echo '# # # It is not necessarily fully secure.                    # # #' . PHP_EOL;
echo '# # # It uses the OUTDATED and UNSECURE algorithm DES.       # # #' . PHP_EOL;
echo '# # # It uses FIXED key that should NEVER used,              # # #' . PHP_EOL;
echo '# # # instead use random generated keys.                     # # #' . PHP_EOL;
echo '# # # DO NOT USE THIS CODE IN PRODUCTION                     # # #' . PHP_EOL;

$plaintext = 'The quick brown fox jumps over the lazy dog';
echo 'plaintext: ' . $plaintext . PHP_EOL;

echo PHP_EOL . 'DES ECB String encryption with fixed key' . PHP_EOL;

// fixed encryption key 8 bytes long
$desKey = '12345678';
echo 'desEncryptionKey (Base64): ' . base64_encode($desKey) . PHP_EOL;

// encryption
echo PHP_EOL . '* * * Encryption * * *' . PHP_EOL;
$desCiphertextBase64 = desEcbEncryptToBase64($desKey, $plaintext);
echo 'ciphertext: ' . $desCiphertextBase64 . PHP_EOL;
echo 'output is (Base64) ciphertext' .PHP_EOL;
echo PHP_EOL . '* * * Decryption * * *' . PHP_EOL;
echo 'ciphertext: ' . $desCiphertextBase64 . PHP_EOL;
echo 'input is (Base64) ciphertext' .PHP_EOL;
$decryptedtext = desEcbDecryptFromBase64($desKey, $desCiphertextBase64);
echo 'plaintext: ' . $decryptedtext . PHP_EOL;

echo PHP_EOL . 'Triple DES ECB String encryption with fixed key (16 bytes)' . PHP_EOL;

// fixed encryption key 16 bytes long
$des2Key = '1234567890123456';
echo 'des2EncryptionKey (Base64): ' . base64_encode($des2Key) . PHP_EOL;

// encryption
echo PHP_EOL . '* * * Encryption * * *' . PHP_EOL;
$des2CiphertextBase64 = des2EcbEncryptToBase64($des2Key, $plaintext);
echo 'ciphertext: ' . $des2CiphertextBase64 . PHP_EOL;
echo 'output is (Base64) ciphertext' .PHP_EOL;
echo PHP_EOL . '* * * Decryption * * *' . PHP_EOL;
echo 'ciphertext: ' . $des2CiphertextBase64 . PHP_EOL;
echo 'input is (Base64) ciphertext' .PHP_EOL;
$decryptedtext = des2EcbDecryptFromBase64($des2Key, $des2CiphertextBase64);
echo 'plaintext: ' . $decryptedtext . PHP_EOL;

echo PHP_EOL . 'Triple DES ECB String encryption with fixed key (24 bytes)' . PHP_EOL;

// fixed encryption key 24 bytes long
$des3Key = '123456789012345678901234';
echo 'des3EncryptionKey (Base64): ' . base64_encode($des3Key) . PHP_EOL;

// encryption
echo PHP_EOL . '* * * Encryption * * *' . PHP_EOL;
$des3CiphertextBase64 = des3EcbEncryptToBase64($des3Key, $plaintext);
echo 'ciphertext: ' . $des3CiphertextBase64 . PHP_EOL;
echo 'output is (Base64) ciphertext' .PHP_EOL;
echo PHP_EOL . '* * * Decryption * * *' . PHP_EOL;
echo 'ciphertext: ' . $des3CiphertextBase64 . PHP_EOL;
echo 'input is (Base64) ciphertext' .PHP_EOL;
$decryptedtext = des3EcbDecryptFromBase64($des3Key, $des3CiphertextBase64);
echo 'plaintext: ' . $decryptedtext . PHP_EOL;

?>
